presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -366,6 +372,112 @@ else if (firstDependencies.stream().allMatch(right::canResolve) && secondDepende
         return new RelationPlan(root, analysis.getScope(node), outputSymbols);
     }
 
+    private RelationPlan planJoinUsing(Join node, RelationPlan left, RelationPlan right)
+    {
+        /* Given: l JOIN r USING (k1, ..., kn)
+
+           produces:
+
+            - project
+                    coalesce(l.k1, r.k1)
+                    ...,
+                    coalesce(l.kn, r.kn)
+                    l.v1,
+                    ...,
+                    l.vn,
+                    r.v1,
+                    ...,
+                    r.vn
+              - join (l.k1 = r.k1 and ... l.kn = r.kn)
+                    - project
+                        cast(l.k1 as commonType(l.k1, r.k1))
+                        ...
+                    - project
+                        cast(rl.k1 as commonType(l.k1, r.k1))
+         */
+
+        List<Identifier> joinColumns = ((JoinUsing) node.getCriteria().get()).getColumns();
+
+        Analysis.JoinUsingAnalysis joinAnalysis = analysis.getJoinUsing(node);
+
+        ImmutableList.Builder<JoinNode.EquiJoinClause> clauses = ImmutableList.builder();
+
+        Map<Identifier, Symbol> leftJoinColumns = new HashMap<>();
+        Map<Identifier, Symbol> rightJoinColumns = new HashMap<>();
+
+        Assignments.Builder leftCoercions = Assignments.builder();
+        Assignments.Builder rightCoercions = Assignments.builder();
+
+        leftCoercions.putIdentities(left.getRoot().getOutputSymbols());
+        rightCoercions.putIdentities(right.getRoot().getOutputSymbols());
+        for (int i = 0; i < joinColumns.size(); i++) {
+            Identifier identifier = joinColumns.get(i);
+            Type type = analysis.getType(identifier);
+
+            // compute the coercion for the field on the left to the common supertype of left & right
+            Symbol leftOutput = symbolAllocator.newSymbol(identifier, type);
+            int leftField = joinAnalysis.getLeftJoinFields().get(i);
+            leftCoercions.put(leftOutput, new Cast(left.getFieldMappings().get(leftField).toSymbolReference(), type.getTypeSignature().toString()));
```

Comment:
What if there is no coercion, see: `QueryPlanner#coerce` (cast is not inserted)

Reply:
It should be harmless, since every type can be cast to itself. But, let me try to avoid adding them if unnecessary.

Reply:
Maybe they are not needed at `QueryPlanner#coerce` too.

Reply:
I haven't seen any rule that removes coercion. I don't think such code exists.

Reply:
It gets ugly, so I'll just leave it as is (i.e., plan it the naive way) and let the optimizer remove them if not necessary. I added a PR for that other issue, so I'll merge this one after that one gets in.

Reply:
I think you should also derive if the cast is type only coercion here and put that information into `Cast`.

Otherwise we could remove `typeOnly` from `Cast` node completely and rely on `TypeManager` in places where it's needed.